SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AutodiffClosureSpecialization.swift

```diff
@@ -0,0 +1,625 @@
+                if usedReabstractionClosures.contains(closureInst) {
+                    continue
+                }
+                
+                var closureInfo: ClosureInfo? = nil
+
+                // Worklist of operands
+                var uses = Array(closureInst.uses)
+
+                // Live range end points.
+                var usePoints: [Instruction] = []
+
+                // Set of possible arguments for mark_dependence. The base of a
+                // mark_dependence we copy must be available in the specialized function.
+                var possibleMarkDependenceBases = ValueSet(self.context)
+                defer {
+                    possibleMarkDependenceBases.deinitialize()
+                }
+
+                if let pai = closureInst as? PartialApplyInst {
+                    for arg in pai.arguments {
+                        possibleMarkDependenceBases.insert(arg)
+                    }
+                }
+
+                var haveUsedReabstraction = false
+
+                // Uses may grow in this loop
+                for useIndex in 0..<uses.count {
+                    let use = uses[useIndex]
+                    usePoints.append(use.instruction)
+
+                    // Recurse through conversions
+                    if let cfi = use.instruction as? ConvertFunctionInst {
+                        uses.append(contentsOf: cfi.uses)
+                        possibleMarkDependenceBases.insert(cfi)
+                        continue
+                    }
+
+                    if let cvt = use.instruction as? ConvertEscapeToNoEscapeInst {
+                        uses.append(contentsOf: cvt.uses)
+                        possibleMarkDependenceBases.insert(cvt)
+                        continue
+                    }
+
+                    // Look through reabstraction thunks
+                    if let pai = use.instruction as? PartialApplyInst {
+                        // Only look at reabstraction thunks if the closure obtained
+                        // as a result of this partial_apply is not the pullback returned
+                        // from this autodiff VJP.
+                        if !isPullbackInResultOfAutodiffVJP(pai: pai) {
+                            // Reabstraction can cause series of partial_apply to be emitted. It
+                            // is okay to treat these like conversion instructions. Current
+                            // restriction: if the partial_apply does not take ownership of its
+                            // argument we don't need to analyze which partial_apply to emit
+                            // release for (its all of them).
+                            let calleeType = pai.callee.type
+                            if pai.isPartialApplyOfReabstractionThunk && isSupportedClosure(pai) && 
+                                calleeType.isNoEscapeFunction && calleeType.isThickFunction {
```

Comment:
formatting: if the condition needs multiple lines, please put the `{` into a separate line (like swift-format would do):
```
  if a &&
     b            // <- align indentation with previous condition `a`
  {
     ...
  }
```

Reply:
Do we use swift-format to format this code? Or do we manually try to conform to swift-format's formatting?

Reply:
I'm specifically asking because it looks like swift-format, formats an if statement with multiple conditions a bit differently from what you're suggesting.

Reply:
We had some ideas to run swift-format, but we don't do it yet. So yes, let's try to match swift-format's formatting rules. Although, there are some little things which swift-format does not well today, like aligning multi-line if-conditions.
If you want to take a look, here is an experiment to run swift-format on the SwiftCompilerSources: https://github.com/apple/swift/pull/67275